Db change aware cache/main

Vivekgupta008 opened the pull request:
## ✅ PR: Enable Change-Aware Context and Embedding Caching

### Summary

This PR addresses the issue where the BabyNest agent **recomputed context and regenerated embeddings** on every query, even when the underlying SQLite database had not changed. This led to unnecessary computation, disk I/O, and degraded performance — especially in offline or resource-constrained environments.

---

### 🔧 What’s Implemented

#### ✅ 1. **Database Hash-Based Staleness Detection**
- The `ContextCache` now compares a **calculated database hash** (using relevant tables: `profile`, `weekly_weight`, etc.) against a stored hash.
- If the hash is unchanged, cached context (memory or disk) is reused.
- If the hash changes, context is rebuilt and stored again.

#### ✅ 2. **Optimized `get_context()` Logic**
- Avoids duplicate calls to `_calculate_db_hash()`.
- Uses a single comparison to detect staleness and decide whether to rebuild context.

#### ✅ 3. **Persistent Cache for Context**
- Cached context is stored in memory and written to disk (`cache/context_default.json`) for reuse across sessions.

#### ✅ 4. **Auto-Trigger Vector Store Refresh**
- On real DB change, the system:
  - Invalidates the context cache
  - Calls `update_vector_store()` to regenerate ChromaDB embeddings

---

### 🧠 Why This Matters

- Prevents unnecessary recomputation on every agent query
- Ensures cache remains valid as long as data is unchanged
- Reduces memory and CPU usage in offline mode
- Maintains synchronization between user data and AI responses

---

### ✅ Benefits

| Before                          | After                                   |
|----------------------------------|------------------------------------------|
| Context rebuilt every time      | Context reused if unchanged              |
| Embeddings always refreshed     | Only refreshed if profile/tracking data changes |
| Unnecessary DB reads + I/O      | Efficient reads only when required       |
| Stale/duplicate cache writes    | Minimal disk activity                    |

---

### 🧪 Tested Scenarios

- [x] Reused context and embeddings when no DB changes
- [x] Triggered regeneration only when data changes
- [x] Accurate agent responses aligned with updated profile/symptoms
- [x] Fast response time after initial cache

---

Fixes : #77 
Please review this @bhavik-mangla 

---

### 📎 Notes

- Observer thread checks DB changes every 2 seconds
- `.state` file tracks last known state of monitored tables
- Prepared the codebase for future multi-user or multi-profile caching logic




<img width="350" height="450" alt="image" src="https://github.com/user-attachments/assets/53acccf9-5e8f-4cf6-8728-ad400507f020" />

<img width="350" height="450" alt="image" src="https://github.com/user-attachments/assets/64b78cb1-490a-4583-907d-63327395267d" />

<img width="700" height="800" alt="image" src="https://github.com/user-attachments/assets/f5212210-dc28-45ef-88a1-22b57732fcce" />



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Context-aware AI agent with per-user persistent cache, event-driven cache updates, vector-backed knowledge, and task recommendations; new endpoints for context, tasks, cache status, stats, and cleanup.
  * Frontend: AgentProvider, Tasks screen with week filtering/prioritization, and model prompt prep for local LLM.

* **Documentation**
  * Detailed backend, cache, integration, and frontend docs added.

* **Tests**
  * Demo, integration, and multiple cache-validation tests added.

* **Chores**
  * Added profile table and updated frontend BASE_URL; default cache fixtures included.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Vivekgupta008 commented on Backend/agent/context.py:
```diff
@@ -1,41 +1,92 @@
-        for a in appointments:
-            result += f"- {a['title']} on {a['appointment_date']} at {a['appointment_time']} ({a['appointment_status']})\n"
+    # Fetch and format appointments
+    appointments = db.execute("SELECT title, appointment_date, appointment_time, appointment_status FROM appointments ORDER BY appointment_date").fetchall()
+    for i, a in enumerate(appointments):
+        doc_content = f"Appointment: {a['title']} on {a['appointment_date']} at {a['appointment_time']} (Status: {a['appointment_status']})"
+        docs.append(doc_content)
+        ids.append(f"appt_{a['id']}")
+        metadatas.append({"source": "appointments"})
 
-        result += "\nWeight Logs:\n"
-        for w in weights:
-            result += f"- Week {w['week_number']}: {w['weight']}kg ({w['note']})\n"
-
-        result += "\nSymptoms:\n"
-        for s in symptoms:
-            result += f"- Week {s['week_number']}: {s['symptom']} ({s['note']})\n"
+    # Fetch and format weight logs
+    weights = db.execute("SELECT week_number, weight, note FROM weekly_weight ORDER BY week_number").fetchall()
+    for i, w in enumerate(weights):
+        doc_content = f"Weight Log Week {w['week_number']}: {w['weight']}kg. Note: {w['note']}"
+        docs.append(doc_content)
+        ids.append(f"weight_{a['id']}")
+        metadatas.append({"source": "weight_logs"})
+    
+    # Fetch and format symptoms
+    symptoms = db.execute("SELECT week_number, symptom, note FROM weekly_symptoms ORDER BY week_number").fetchall()
+    for i, s in enumerate(symptoms):
+        doc_content = f"Symptom Week {s['week_number']}: {s['symptom']}. Note: {s['note']}"
+        docs.append(doc_content)
+        ids.append(f"symptom_{a['id']}")
+        metadatas.append({"source": "symptoms"})
+        
+    return docs, ids, metadatas
```
There are various reasons I went with the above approach: 
- Setting db.row_factory = sqlite3.Row modifies the connection’s behavior for all subsequent queries using that connection
- Accessing columns by name (dictionary style) can be slightly slower than tuple indexing, especially if large datasets are fetched repeatedly
- The original queries might have omitted id if it was not needed for the particular processing or indexing

Vivekgupta008 replied:
Thanks for the fix! That definitely solves the bugs.
For BabyNest though, since it’s running fully offline on mobile, I want to keep things simple and lightweight. Changing the whole connection’s row factory feels like it might cause unexpected issues elsewhere, especially as the app grows.
Using tuple indexes feels clearer and safer to me — I know exactly what data I’m pulling without relying on extra SQLite features. Plus, it keeps the code straightforward, which is important when handling sensitive pregnancy info.

So, while your approach works, I think sticking to tuple indexing strikes a better balance for BabyNest’s performance and stability.

bhavik-mangla commented on Backend/agent/cache.py:
```diff
@@ -0,0 +1,285 @@
+                conn.close()
+    
+    def get_context(self, user_id: str = "default") -> Optional[Dict[str, Any]]:
+        """Get user context from cache or build it if needed."""
```
Get user context from cache only, unless it's null which should be handled in onboarding of user

bhavik-mangla commented on Backend/app.py:
```diff
@@ -42,13 +47,41 @@ def run_agent():
+        return jsonify({"error": str(e)}), 500
+
+@app.route("/agent/cache/status", methods=["GET"])
+def get_cache_status():
```
Its really important to clean your code
Include only the necessary functions
Its really hard for someone to go through so many functions which almost do similar tasks

bhavik-mangla commented on Backend/app.py:
```diff
@@ -42,13 +47,41 @@ def run_agent():
+    try:
+        user_context = agent.get_user_context()
+        return jsonify({
+            "cache_status": "active",
```
include current date when context is passed to llm

bhavik-mangla commented on Backend/agent/db_observer.py:
```diff
@@ -0,0 +1,152 @@
+import sqlite3
```
We dont need a database observer, as we plan to update cache everytime we make changes to db
Remove this class

Vivekgupta008 commented:
Have added update_cache and removed invalidate_cache as discussed, also made separate test files for testing individual feature independently. Also added cache size as shown below 
<img width="757" height="145" alt="image" src="https://github.com/user-attachments/assets/e9bc113c-907f-4a5b-bc09-7fe2b9d6690e" />

Vivekgupta008 commented:
Below are the screenshots of all the tests file working just fine for reference

<img width="562" height="869" alt="image" src="https://github.com/user-attachments/assets/e0d9900a-3c42-4a0c-a396-36b27deaaf73" />



<img width="300" height="911" alt="image" src="https://github.com/user-attachments/assets/6aea4472-7df7-4310-bcc6-fc7b0356be1f" />



<img width="455" height="283" alt="image" src="https://github.com/user-attachments/assets/b83808da-036d-415d-8dbf-5e3b67ca3bac" />



<img width="468" height="382" alt="image" src="https://github.com/user-attachments/assets/5b7ac8db-a082-4df3-a0d7-40373bf79fe8" />


Please have a look @bhavik-mangla

bhavik-mangla commented:
Kudos @Vivekgupta008 !! ☺️☺️
